### PR TITLE
docs: Add a LaTeX equivalent of the GameCube DSP User's Manual

### PR DESCRIPTION
**Pull request by lioncash**
The existing manual is nice to read and get up to speed on things, however it has one major problem. It's actually a pain to extend or modify, despite being released under the GNU documentation license: It's in PDF format.

I've taken the time to reproduce what the document more-or-less looks like in order to allow extending the document using open-source tools while also using a more general solution like LaTeX, which can generate the document into multiple formats if desirable.

This should make extending the document much easier (and not lock the information into one concrete format).

While the formatting could use some minor touching up, I did my best to adhere it to the existing formatting as much as possible. An example of what it currently looks like can be found [here](https://drive.google.com/file/d/1LBZu6YGsDdvJFQGDJmUvInXARwPZR-m3/view?usp=sharing)

### CODE REVIEW

**Comment by BhaaLseN** [comment]
Nice! I still got a modified PDF Version lying around with updated tables/flags; I'll look into incorporating those changes once this is merged. I too tried to make something writable of it, but I turned to XML instead (and miserably failed). The obvious choice (straight TeX) didn't occur to me back then.

File: docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex [hunk 1]
```diff
@@ -0,0 +1,3737 @@
...
+  \item \Register{\$st3} -- Loop counter register
+\end{itemize}
+
+Stacks are implemented in hardware and have limited depth. The data stack is limited to four values
+and the call stack is limited to eight values. The loop stack is limited to four values.
+Upon underflow or overflow of any of the stack registers exception \Exception{STOVF} is raised.
+
+The loop stack is used to control execution of repeated blocks of instructions. Whenever there is a value in
+\Register{\$st2} and the current PC is equal to the value in \Register{\$st2}, then the value in \Register{\$st3} is decremented.
+If the value is not zero, then the PC is modified by the value from call stack \Register{\$st0}. Otherwise values from callstack
```
Comment (Tilka) [inline docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex]: callstack -> call stack

File: docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex [hunk 2]
```diff
@@ -0,0 +1,3737 @@
...
+\Register{\$16} & \Register{\$r10} & \Register{\$ac0.h}   &                       \\ \hline
+\Register{\$17} & \Register{\$r11} & \Register{\$ac1.h}   &                       \\ \hline
+\Register{\$18} & \Register{\$r12} & \Register{\$config}  &                       \\ \hline
+\Register{\$19} & \Register{\$r13} & \Register{\$sr}      &                       \\ \hline
+\Register{\$20} & \Register{\$r14} & \Register{\$prod.l}  &                       \\ \hline
+\Register{\$21} & \Register{\$r15} & \Register{\$prod.m1} &                       \\ \hline
+\Register{\$22} & \Register{\$r16} & \Register{\$prod.h}  &                       \\ \hline
+\Register{\$23} & \Register{\$r17} & \Register{\$prod.m2} &                       \\ \hline
+\Register{\$24} & \Register{\$r18} & \Register{\$ax0.l}   &                       \\ \hline
+\Register{\$25} & \Register{\$r19} & \Register{\$ax1.1}   &                       \\ \hline
```
Comment (Tilka) [inline docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex]: `$ax1.1` should be `$ax1.l` (lowercase L)

File: docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex [hunk 3]
```diff
@@ -0,0 +1,3737 @@
...
+\Register{\$26} & \Register{\$r1A} & \Register{\$ax1.h}   &                       \\ \hline
+\Register{\$27} & \Register{\$r1B} & \Register{\$ax1.h}   &                       \\ \hline
```
Comment (Tilka) [inline docs/DSP/GameCube_DSP_Users_Manual/GameCube_DSP_Users_Manual.tex]: The duplicate `$ax1.h` seems to be a typo that's also present in the original. I'm guessing the first one should be `$ax0.h`.

**Comment by shuffle2** [comment]
I'm kind of surprised that github doesn't seem to have a latex preview builtin :'(

**Comment by lioncash** [comment]
@Tilka Amended the tex file

**Comment by BhaaLseN** [comment]
Just noticed, did you intentionally just move the old PDF and _not_ include the newly generated one?

**Comment by lioncash** [comment]
Initially it was, but I intended to replace it with the newer one when making the PR, good catch

**Comment by lioncash** [comment]
Actually, would it be preferable to only provide the tex file? Like @BhaaLseN stated in #6641, multiple PDF updates would just add yet another copy of the binary file into the repository history, which kind of sucks.

**Comment by BhaaLseN** [comment]
It has both advantages (no binary updates) and disadvatanges (you need a LaTeX processor to transform it before you can read it). It probably doesn't matter too much if its updated every once in a blue moon; but might blow up repo size a lot if it happens more frequently.

Can't really tell which one is the more practical for "end-users" (folks that want to read the PDF); since it's a dev doc chances are the reader at least has an idea of how to make their own PDF.

**Comment by lioncash** [comment]
Opted to only provide the tex file in the repository, given it's a dev-oriented document.

**Comment by sepalani** [comment]
TBH, even if it's a dev-oriented doc I don't think it's a valid reason to force people to install yet another tool/dependency. Especially, since its a documentation and not even a tool. I mean, devs won't expect to install LaTeX to be able to read the DSP doc. What if people want to read it on their phone or whatever device which might not support LaTeX but might be suitable to read PDFs/docs.

Plus, I found LaTeX to be a pain regarding dependency install. I didn't manage to compile this file on all my systems with LaTeX so far, Windows and Linux. So, it doesn't compile right off the bat. Unless, there is a tutorial to install all its dependencies, I don't think that's nice for people who want to start to work on DSP not to provide a ready to use compiled version of the documentation.

I won't be against removing the PDF version as long as a compiled/PDFs version is available. For instance, it can be stored using GitHub Git LFS or attached to a PR and the link shared through a README.md file in the appropriate folder.

**Comment by BhaaLseN** [comment]
[MiKTeX](https://miktex.org/) seemed fairly painless to me, but yeah, I kinda agree there too. Anything other than LFS is more of a pain to maintain, so I'd advise against that.

Whether we decide to use LFS and commit the PDF in the end or stick to the bare LaTeX (or perhaps a 3rd option with a different format) - I have no particular preference to either.

**Comment by lioncash** [comment]
@sepalani I won't be putting an effectively live binary file into the repo -- end of story in that regard. 

> TBH, even if it's a dev-oriented doc I don't think it's a valid reason to force people to install yet another tool/dependency. 

Any in-the-middle format essentially requires this. Even basic things like Doxygen-compatible comments still actually require Doxygen to make the formalized documentation. So I would say this isn't a very good argument.

> Plus, I found LaTeX to be a pain regarding dependency install

Not being used to something is not a good reason to not use something. I've been able to compile it both on Linux and Windows. Through the package manager, install `memoir` and any of the packages that aren't already present (which can be found by looking at the `\usepackage{}` commands at the top of the file. You've also not posted any of the issues you're actually having.

**Comment by sepalani** [comment]
> I won't be putting an effectively live binary file into the repo -- end of story in that regard.

I do agree, that's why I suggested to [attach the PDF to one of the PR's comments](https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/) and share its link through a `README.md`. For instance, this is the original document: [GC DSP.pdf](https://github.com/dolphin-emu/dolphin/files/1911706/GC.DSP.pdf). Plus, Dolphin doesn't use the GitHub's issue system so it's quite unlikely to hit some kind of quota if there is a hidden one.

> Any in-the-middle format essentially requires this. Even basic things like Doxygen-compatible comments still actually require Doxygen to make the formalized documentation. So I would say this isn't a very good argument.

That's a bad counter argument, most people are exposing the generated documentation. That way users don't have to generate it by themselves. I don't need to install Doxygen if I want to learn `mbedtls`. However, it's highly encouraged to install it if I want to update its generated documentation. So, there are two different targets with different needs.

> This is very much a "not my problem" scenario, sorry. I could apply a lot of "what if I wanted to read on [x]" scenarios to a lot of other things as well.

Users might be facing specific issues because the version of LaTeX they're using or whatever tool they're using to compile it is broken. Not providing a ready to use version of the documentation is counter-productive and will most likely make them complain for a valid reason.

> Not being used to something is not a good reason to not use something. I've been able to compile it both on Linux and Windows. Through the package manager, install memoir and any of the packages that aren't already present (which can be found by looking at the \usepackage{} commands at the top of the file), then run pdflatex. You've also not posted any of the issues you're actually having.

I haven't posted the issue I'm facing because I'm familiar with LaTeX and I was able to solve it. But still, as I said, it's really unpleasant to provide something that should work without explicitly telling which are the dependencies and what is the recommended way to install them. And might be the source of people creating issues because "it doesn't work™". Plus, initially, the user goal was to learn how DSP work, not how to compile a doc with LaTeX (even if that's a good opportunity to learn LaTeX).

**Comment by lioncash** [comment]
Yeah, I think that sounds fine, but it should likely go somewhere more permanent (e.g. being uploaded to the wiki, or some other dedicated place). If for whatever reason in the future, we migrate off Github, storing them in attachments will be a problem. I've made an account on the wiki to upload it, however I'm still waiting out the autoconfirmation period to pass so I can actually upload things.

The rest of this comment isn't necessary, but for the sake of addressing them:

> That's a bad counter argument...

Fair

> Users might be facing specific issues because the version of LaTeX they're using or whatever tool they're using to compile it is broken

Yeah, providing one pre-built is indeed convenient (and I never said we shouldn't do that, just it won't be in the repository). Even without that aspect I still consider it a not-my-problem case. It's not our job to worry about environmental details that aren't under our control or outside of the requirements we state. If someone's LaTeX installation is broken and they want to build the LaTeX document, then they're out of luck in general. Just like compiling an actual codebase's code is not guaranteed to work if they use a compiler under the requirements or if their compiler is fundamentally broken somehow. This is a complaint point to direct to the maintainer/provider of the LaTeX installation on [x] platform. Not providing a usable PDF is valid; "I'm using broken/incompatible software to build [thing] and it doesn't work" is absolutely not.

> without explicitly telling which are the dependencies

The readme explicitly tells you the dependencies are the packages specified by `\usepackage` commands. Like yeah, I could duplicate a bulleted list (and now have two lists to update instead of one when changing dependencies). If looking at `\usepackage{thing}` is difficult, then there was a problem before there was even an attempt at setting up the environment.

> what is the recommended way to install them. 

It's also not our job to explain how a package manager works or how to fundamentally use LaTeX. If it's really not obvious from "install" to mean "just install it in a way that's usable by LaTeX" (i.e. what is meant 99% of the time install is used in a software context when referring to [some thing] being used with another thing), I'm placing the issue in the realm of PEBCAK. We can point to https://en.wikibooks.org/wiki/LaTeX/Installing_Extra_Packages if it's *really* that ambiguous, but aside from that, I think this borders on unfair hairsplitting for a particular argument issue, considering most guides on a particular thing outside of beginner guides assume basic working knowledge of the tools they're using or the ability of the person reading the guide to be able to gain that working knowledge by reading the actual tool's documentation.

**Comment by BhaaLseN** [comment]
Maybe we can pester @delroth to watch for .tex changes on master and build+upload it automatically ;)
Linking is then done with a dolp.in link that always points to the most recent one.

**Comment by delroth** [comment]
@BhaaLseN should have written it in Sphinx/ReST instead of "raw" LaTeX and then we could have just used Readthedocs :)

**Comment by BhaaLseN** [comment]
Still means we need a build process _somewhere_ :)

**Comment by lioncash** [comment]
ReST would have actually been significantly more annoying for doing stylistic things like adding bitfield notation around instruction opcodes, etc, without repeating yourself over and over again (or literally for formatting opcodes when it comes to sectioning, which LaTeX removes with environments and commands) which is why I didn't go with it, for what it's worth.

**Comment by degasus** [comment]
It took me 2 minutes to copy the source file to eg overleaf and to get a preview online without installing anything. So I'm very fine with not having the pdf in our repository.

**Pull request merged**